### ISSUE TITLE
Whether to consider adding a Chinese document?

### ISSUE THREAD

**Issue by jweboy** [fastify/fastify#722]
I'm using Fastify, but at the moment there is only an English document, which is not very convenient for someone who is not good at English, so I translated it with Google on a Chinese translation. Would you suggest adding a Chinese document or I would like to contribute a Chinese document for you. In conjunction with Google Translate, I will speak more semantic and straightforward about document translation so that when I am studying, I am able to help more developers....

**Comment by jsumners** [comment]
Are you suggesting use Google Translate for the initial translation followed up by you modifying that translation to be correct? Automatic translations are usually quite terrible.

**Comment by jweboy** [comment]
Maybe I did not express clearly. My awareness is assisted by Google Translate, through my personal translator and interpreter, which makes my document translation more understandable and understandable, making it easier for developers looking at Chinese documents to get started quickly.

**Comment by nwoltman** [comment]
There could be a folder for each translated language inside the `docs` folder (like `docs/zn-cn` and `docs/it`). @lmammino Could this be integrated into the website easily?

**Comment by fralonra** [comment]
Many developers in China dont't even want to try Fastify for there is English documents only. 
So I suggest adding translations for other languages not only Chinese.

I can help translate the documents into Chinese if translation is needed.

**Comment by delvedor** [comment]
Hi @fralonra, thank you for stepping in!
If you want to translate our documentation in Chinese would be awesome, but it requires a big effort because you should keep it in sync with all future changes.
Since no one of our team knows Chinese, we can’t guarantee that the documentation will always be in sync and from my point of view it is better one thing done well than two done badly.

Do you want to take this effort?

**Comment by vincent178** [comment]
I can help as well.

**Comment by mcollina** [comment]
I will set a repo for this doc effort and add both of you as members. Is that ok?

**Comment by vincent178** [comment]
sure, thanks @mcollina

**Comment by fralonra** [comment]
ok, thanks @mcollina

**Comment by delvedor** [comment]
Thank you folks! :)

**Comment by mcollina** [comment]
I've created the repo at https://github.com/fastify/docs-chinese. I've sent you an invite for a team that has admin rights on the repo!

Let us know when it's populated. We can add a link into our docs to that repo and close this issue then.

**Comment by lmammino** [comment]
That's awesome! I love the idea of being able to reach a wider world wide community and the challenge of figuring out what would be the best way to deal with multi-language documentation.

Maybe we should also start to figure out how to integrate this with the current website.

It would be very nice to have the website support multiple languages, but i am not sure just yet how to do so.

Just a couple of thoughts to spark a conversation and get some suggestions:

- with the current setup it's very nice that docs and code are kept in the same repo. This enforces consistency because any code change will have to bring doc updates too in order to be considered mergeable. This happens in the same PR, hence there's no coordination needed across repos regarding reviews, versioning and merge.

- Chinese is not our main language and, i believe, none of us thought about supporting multi-language documentation so far. The big risk is to have some translations becoming stale as it's going to be very hard for a PR to include all the translations of the doc before we can proceed with the merge. In this sense i believe having a separate repo might be a good idea, but at the same time it feels like we are giving up on trying to have a consistent multi-language documentation.

An idea, which i am not particularly fond of, could be to have "unofficial" translations, so that there's no implicit promise of consistency and the website might simply link to those without integrating them directly in the build process.

In case we really want to have multi-language docs integrated with the current website, i have few ideas on how to achieve that, but i'd have to run some experiment to see how feasible (and easy to implement) they are.

I look forward to knowing what you think!

**Comment by StarpTech** [comment]
I use https://github.com/facebook/Docusaurus for all my opensource projects. The documentation is generated based on the markdown files too. It supports:

* Versioning
* Localization (with [crowdin](https://crowdin.com/) support)
* Blog posts
* One-liner to integrate free Algolia search
* Customized via React

I like it very much and I think it could save us much time to standarzize things and to improve the overall experience. I like the current website as well so I'm open.

**Comment by lmammino** [comment]
@StarpTech, I will have to have a look at it, never used it before. If we are going to switch only because of the localization support, let's make sure that it works in a way that it is convenient for us.

Of course, if you have any experience with this particular feature I'd be happy to have some concrete advice on how we should organise the files/repos to adopt it at best.

**Comment by mcollina** [comment]
I'm somewhat reluctant in having a different infrastructure. I really like the current one, and before committing to i18n we need to have a good case for those.

I think we cannot commit to multi-language docs at this point, there are too few of us. If some members of our community want to help, having a separate repo would be ok for now. I know there is a problem of stale docs, but shipping multi-lang docs is __hard__.

**Comment by lmammino** [comment]
Good point @mcollina. Let's see what happens with this new translation repo and then we can decide whether it makes sense to have an integrated localised website or not.

**Comment by StarpTech** [comment]
@lmammino It's not only about localization it provides a complete package. In the current state, we will face with other problems too.

* [ ] How to version the whole website?
* [ ] Do we want to provide blog posts on our website?
* [ ] How to integrate a search?
* [ ] _How to localize in multiple languages?_
* [ ] How to ensure structure and maintainability when the website get bigger or more complex? (React could be a good choice)

From my point of view it seems that we will reimplement many things. @mcollina before you disagree please have a look on some examples. Of course, it will take some to migrate but we are smart people and I think this can be done in a week.

> Of course, if you have any experience with this particular feature I'd be happy to have some concrete advice on how we should organise the files/repos to adopt it at best.

@lmammino the [default](https://docusaurus.io/docs/en/site-preparation.html) strategy is to maintain a `/docs` folder where all markdown are listed. The website is stored in the main repository `/website`. As soon as you make any changes to the docs you can generate and publish it with:

```
cd /website
npm run version 1.2.0
npm run publish-gh-pages
```

The website is published via github pages in the `gh-pages` branch.

The ui components are implemented in React.

Some other good points:

*  Sidebar and page informations are managed inside the markdown file as well.
*  No seperate template language everyhthing is react or markdown.

**Comment by mcollina** [comment]
@StarpTech the main advantage of our current system is that it enables us to have the documentation of multiple version of our docs published at the same time. I do not see this possible with docusaurus. More importantly, I do not see how this would be possible with the website living in this repo.

**Comment by StarpTech** [comment]
> @StarpTech the main advantage of our current system is that it enables us to have the documentation of multiple version of our docs published at the same time. I do not see this possible with docusaurus.

Could you elaborate this? Docusaurus also allows to publish multiple versions, rename versions, update versions, provide `/next/en` param to point to the current state and so on.

References:
* https://docusaurus.io/docs/en/versioning.html

**Comment by delvedor** [comment]
@mcollina I think it is better to create a `docs` repo with all the languages inside divided per folders, in this way we will more future proof and maybe it will be easier build the website (@lmammino?).

**Comment by mcollina** [comment]
@delvedor I'm -1 until the Chinese docs are done. I fear about them getting out of sync.

**Comment by delvedor** [comment]
@mcollina what I meant is to create a brand new repository named `docs` or `documentation`, and inside add a folder for every language. Not inside the main Fastify repository (where we will continue to support just english docs).

**Comment by lmammino** [comment]
@mcollina I think docosaurus supports multiple versions of docs, but I haven't played with it so far so I am not sure how to achieve that and, most importantly, if it makes sense to spend our energies in switching build system right now.

Regarding how to manage multiple languages i agree with the idea of having a separate repo for now and see what happens.

For now we can just put a link (maybe above the footer) in our website saying something like:

> Are you looking for translations of this documentation? Check the repo with the unofficial translations.

If, at a later stage, we get enough traction and translate docs are reasonably kept in sync, then we should revisit this discussion.

**Comment by mcollina** [comment]
The approach docosaurous is taking requires we heavily change how docs are built. We would need to go back and publish the docs _for every version_ of Fastify that we have published so far. I'm not very comfortable in doing that change right now, but we might do it in the future if wee need more of these features.

**Comment by StarpTech** [comment]
> The approach docosaurous is taking requires we heavily change how docs are built. We would need to go back and publish the docs for every version of Fastify that we have published so far. I'm not very comfortable in doing that change right now, but we might do it in the future if wee need more of these features.

No! Please check it. In the current state, we have to download all releases and build the docs. In docosaurous there is a `versioned_docs` folder with all previous versions in markdown (Only changes are tracked in a new docs release). The migration process would be not so hard and can be automated. Therefore it allows us to keep docs much easier in sync, you don't have to make a git release for each update (typo, improvements not-api changes).

**Comment by mcollina** [comment]
The current approach is deliberate. We do not want any docs to be published if they are not part of a release.

I’ll check out docusaurus as soon as I can (likely not soon, what we have works).

**Comment by StarpTech** [comment]
Allright, I just want to ensure that we all on the same page :)

**Comment by delvedor** [comment]
Personally I really like our current system, @lmammino did a great job and it is working perfectly, I don't see a reason to change it :)

**Comment by vincent178** [comment]
Hi all,

As the translations are going on, to keep doc in sync, it would be great to at-mention @fralonra and me when there's a doc change.

The other proposal is that, we do translation when there's a big release.

When do you guys think?

Thanks

**Comment by vincent178** [comment]
@fastify/docs-chinese has completed, would it be ok to add a link in @fastify/fastify navigate to @fastify/docs-chinese ?

**Comment by mcollina** [comment]
definitely. Send a PR. I'd also recommend an edit to https://www.fastify.io/!

**Comment by mcollina** [comment]
I think this is done. 🎉
